### PR TITLE
Add QueryGroup API documentation

### PR DESCRIPTION
**Pull request by ruai0511**
This change introduces the QueryGroup CRUD API documentation which we will use to enforce node level resiliency as part of this RFC: https://github.com/opensearch-project/OpenSearch/issues/12342. The QueryGroup schema is in this PR: https://github.com/opensearch-project/OpenSearch/pull/13669

**Create API:**
Request:
```
curl -XPUT "localhost:9200/_query_group" -H 'Content-Type: application/json' -d ' 
{
     "name": "analytics",
     "jvm":0.4,
     "resiliency_mode": "monitor"
}'
```
Response:
```
{"name":"an1alytics","jvm":0.4,"resiliency_mode":"monitor","updatedAt":1718922703953}
```

**Update API:**
Request:
```
curl -XPUT "localhost:9200/_query_group/analytics" -H 'Content-Type: application/json' -d ' 
{
     "jvm":0.3,
     "resiliency_mode": "monitor"
}'
```
Response:
```
{"name":"analytics","jvm":0.3,"resiliency_mode":"monitor","updatedAt":1718922977093}
```
**Get API:**
Request:
```
curl -XGET "localhost:9200/_query_group" //get all QueryGroups
curl -XGET "localhost:9200/_query_group/analytics" //get one QueryGroup
```
Response:
```
{"query_groups":[{"name":"analytics","resiliency_mode":"monitor","jvm":0.3"updatedAt":1718922344489}]}
```

**Delete API:**
Request:
```
curl -XDELETE "localhost:9200/_query_group" //delete all QueryGroups
curl -XDELETE "localhost:9200/_query_group/analytics" //delete one QueryGroup
```
Response:
```
{"deleted":[{"name":"analytics","resiliency_mode":"monitor","jvm":0.3"updatedAt":1718922344489}]}
```

### CODE REVIEW

**Comment by dblock** [comment]
Great start! Iterate to 🟢 . Make sure to write a test story in https://github.com/opensearch-project/opensearch-api-specification/tree/main/tests.

**Comment by ruai0511** [comment]
> Great start! Iterate to 🟢 . Make sure to write a test story in https://github.com/opensearch-project/opensearch-api-specification/tree/main/tests.

Thanks! I have 2 questions
1. I saw this error message. Basically it says I need to refer to the same requestBody for path `PUT /_query_group `and `PUT /_query_group/{name}`. But we have different schemas for those 2 api paths. Is there a way to have different reference for requestBody for them?
```
{
  file: 'namespaces/query_group.yaml',
  location: 'Operation: PUT /_query_group',
  message: "The requestBody must be a reference object to '#/components/requestBodies/query_group.put'."
}
{
Error: [ERROR] 
Total errors:2
  file: 'namespaces/query_group.yaml',
  location: 'Operation: PUT /_query_group/{name}',
  message: "The requestBody must be a reference object to '#/components/requestBodies/query_group.put'."
}
```
3. Where should I create the test story? Is it okay to create another folder under tests called query_group and put my tests in there?

**Comment by dblock** [comment]
For (1) I am not sure, but maybe try using a different operation group? @nhtruong will know for sure.

For (2), this is something under `_query_group` so make a new `query_group` folder.

**Comment by nhtruong** [comment]
@ruai0511 If 2 operations have different behaviors (having different inputs for example), then they should belong to different operation groups. In this case they can be `query_group.create` and `query_group.update`.

File: spec/namespaces/query_group.yaml [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+      description: ''
+      content:
+        application/json:
+          schema:
+            type: object
+            properties:
+              query_groups:
+                type: array
+                items:
+                  $ref: '../schemas/query_group._common.yaml#/components/schemas/QueryGroupResponseStructure'
```
Comment (nhtruong) [inline spec/namespaces/query_group.yaml:113]: Is the word Structure coming from Smithy? It doesn't add any info and should be removed. These names will be generated in some clients. So we're want to keep these short and simple.

File: spec/namespaces/query_group.yaml [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+      description: ''
+      content:
+        application/json:
+          schema:
+            type: object
+            properties:
+              deleted:
+                type: array
+                items:
+                  $ref: '../schemas/query_group._common.yaml#/components/schemas/QueryGroupResponseStructure'
```
Comment (nhtruong) [inline spec/namespaces/query_group.yaml:102]: I notice that all operations have the exact response schema. Is this how it actually works or just a WIP placeholder>

Comment (ruai0511) [reply spec/namespaces/query_group.yaml:102]: They are not exactly the same, the exact response schema is in the description of this PR.

File: spec/namespaces/query_group.yaml [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+          $ref: '#/components/responses/query_group.delete@200'
+components:
+  requestBodies:
+    query_group.create:
+      content:
+        application/json:
+          schema:
+            type: object
+            properties:
+              $ref: '../schemas/query_group._common.yaml#/components/schemas/QueryGroupCreate'
```
Comment (nhtruong) [inline spec/namespaces/query_group.yaml:75]: I think you meant:
```yaml
schema:
   $ref: '../schemas/query_group._common.yaml#/components/schemas/QueryGroupCreate'
```

File: spec/namespaces/query_group.yaml [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+      required: true
+    query_group.update:
+      content:
+        application/json:
+          schema:
+            type: object
+            properties:
+              $ref: '../schemas/query_group._common.yaml#/components/schemas/QueryGroupUpdate'
```
Comment (nhtruong) [inline spec/namespaces/query_group.yaml:83]: Here as well

**Comment by dblock** [comment]
I opened https://github.com/opensearch-project/opensearch-api-specification/issues/398 to add support to run tests against the next version of OpenSearch.

**Comment by dblock** [comment]
With https://github.com/opensearch-project/opensearch-api-specification/pull/410 you can now run tests against the next version of OpenSearch.

**Comment by dblock** [comment]
This was finished in https://github.com/opensearch-project/opensearch-api-specification/pull/649, closing.

@ruai0511 double check that everything in this PR made it there

**Pull request closed without merging**